compiler/src/main/scala/org/scalajs/nscplugin/PrepJSInterop.scala

```diff
@@ -996,6 +996,23 @@ abstract class PrepJSInterop[G <: Global with Singleton](val global: G)
                     "A non-native JS class cannot declare a concrete method " +
                     "named `apply` without `@JSName`")
               }
+            } else {
+              /* Allow an abstract `def apply` only if the owner is a plausible
+               * JS function SAM trait.
+               */
+              val owner = sym.owner
+              val isPlausibleJSFunctionType = {
+                owner.isTrait &&
+                owner.superClass == JSFunctionClass &&
+                samOf(owner.toTypeConstructor) == sym
+              }
+
+              if (!isPlausibleJSFunctionType) {
+                reporter.error(sym.pos,
+                    "A non-native JS type can only declare an abstract " +
+                    "method named `apply` without `@JSName` if it is the "+
+                    "SAM of a trait that extends js.Function")
+              }
```

Comment:
I think we should keep this. One comment: Consider putting the `if` as a `case` guard to reduce nesting.